openmicroscopy/bioformats
# Automated tests: add support for external configuration directory

Currently automated tests are relying upon configuration files (`.bioformats`) being present under the same tree as the data repository. The goal of this infrastructure PR is to remove this requirement and allow external configuration tree to bee used to test the data repository. Ultimately, this should simplify our workflow when modifying readers and configuration files /cc @jburel

To test this PR:

- check the overall set of changes is sensible. Some refactoring in the test classes to use helper methods and hide parameters of type `${property}` set by `ant`
- check the existing test logic is untouched without specifying a configuration directory
  ```
ant test-config -Dtestng.directory=/path/to/data  # should create/append .bioformats files in-place
ant test-automated -Dtestng.directory=/path/to/data  # should test using local .bioformats files
```
- test the configuration files can be generated under a custom directory using `test-config`
  ```
ant test-config -Dtestng.directory=/path/to/data -Dtestng.configDirectory=/path/to/config 
```
Check that for each section, the entries of `.bioformats` file should now be sorted by key. This is meant to simplify diff examination when updating these files.
- modify the created `.bioformats` file under `/path/to/config` and test the new ant variable picks up this configuration file instead of the local file under the data directory tree using
  ```
ant test-automated -Dtestng.directory=/path/to/data -Dtestng.configDirectory=/path/to/config
```

Other scenarios that can be tested:
- test a more complex tree (multiple level of hierarchies under `/path/to/data`)
- test multiple data folders
- test removing the `.bioformats` files under the data directory with the `configDirectory` property set does not affect the tests (for future data repo cleanup)
